src/lib/utils/lib.ts

```diff
@@ -56,108 +57,125 @@ export const getTransactionReceipt = async (
-    let result
-    let start = startArbBlock
-    let end = endArbBlock
+  while (start <= end) {
+    // Calculate the midpoint of the current range.
+    const mid = start + Math.floor((end - start) / 2)
 
-    while (start <= end) {
-      // Calculate the midpoint of the current range.
-      const mid = start + Math.floor((end - start) / 2)
+    const l1Block = await getL1Block(mid)
 
-      const l1Block = await getL1Block(mid)
-
-      // If the midpoint matches the target, we've found a match.
-      // Adjust the range to search for the last occurrence.
-      if (l1Block === targetL1BlockNumber) {
-        result = mid
-        start = mid + 1
-      } else if (l1Block < targetL1BlockNumber) {
-        // If the L1 block number is less than the target, adjust the range to the upper half.
+    // If the midpoint matches the target, we've found a match.
+    // Adjust the range to search for the first or last occurrence.
+    if (l1Block === forL1Block) {
+      if (allowGreater) {
         start = mid + 1
       } else {
-        // If the L1 block number is greater than the target, adjust the range to the lower half.
         end = mid - 1
       }
+    } else if (l1Block < forL1Block) {
+      start = mid + 1
+    } else {
+      end = mid - 1
     }
 
-    if (typeof result === 'undefined') {
-      throw new Error(`No L2 range found for L1 block: ${targetL1BlockNumber}`)
-    }
+    // Stores last valid L2 block correlating to the current L1 block.
+    // We store the L1 block too and return them as a pair.
+    if (l1Block) {
+      // Store lesser or greater blocks only if there is no result for the block of interest.
+      const shouldStoreLesser =
+        !allowGreater && l1Block < forL1Block && resultForL1Block !== forL1Block
 
-    return BigNumber.from(result)
-  }
+      const shouldStoreGreater =
+        allowGreater && l1Block > forL1Block && resultForL1Block !== forL1Block
 
-  // Adjust the range to ensure it encompasses the target L1 block number.
-  // We lower the range in increments if the start of the range exceeds the L1 block number.
-  while (
-    (await getL1Block(startArbBlock)) > targetL1BlockNumber &&
-    startArbBlock >= 1
-  ) {
-    // Lowering the range.
-    endArbBlock = startArbBlock
-    startArbBlock = Math.max(
-      1,
-      Math.floor(startArbBlock - currentArbBlock * 0.2)
-    )
+      if (l1Block === forL1Block || shouldStoreLesser || shouldStoreGreater) {
+        lastValidL2Block = mid
+        resultForL1Block = l1Block
+      }
+    }
   }
 
-  return await Promise.all([getL2StartBlock(), getL2EndBlock()])
+  return { l2Block: lastValidL2Block, forL1Block: resultForL1Block }
+}
+
+export const getBlockRangesForL1Block = async (props: {
+  forL1Block: number
+  provider: JsonRpcProvider
+}) => {
+  return await Promise.all([
+    getFirstBlockForL1Block({ ...props, allowGreater: false }),
+    getFirstBlockForL1Block({ ...props, allowGreater: true }),
```

Comment:
I haven't really looked at the logic yet, but if implemented as what I expected this should be
```suggestion
    getFirstBlockForL1Block({ props.forL1Block + 1, props.provider, allowGreater: true }) - 1,
```